### PR TITLE
Simplify unification variables in error messages

### PR DESCRIPTION
**Pull request by balajirrao**
Unification variables in error messages are renumbered to make
them more friendly.

### CODE REVIEW

File: src/Language/PureScript/Errors.hs [hunk 1]
```diff
@@ -514,32 +573,37 @@ prettyPrintSingleError full e = prettyPrintErrorMessage (if full then e else sim
...
-prettyPrintMultipleErrorsWith :: String -> String -> Bool -> MultipleErrors -> String
-prettyPrintMultipleErrorsWith intro _ full  (MultipleErrors [e]) = renderBox $
-  Box.vcat Box.left [ Box.text intro
-                    , prettyPrintSingleError full e
-                    ]
-prettyPrintMultipleErrorsWith _ intro full  (MultipleErrors es) = renderBox $
-  Box.vcat Box.left [ Box.text intro
-                    , Box.vsep 1 Box.left $ map (Box.moveRight 2 . prettyPrintSingleError full) es
-                    ]
+prettyPrintMultipleWarnings :: Bool -> MultipleErrors -> State [Unknown] String
```
Comment (paf31) [inline src/Language/PureScript/Errors.hs]: You have quite a few calls to `flip L.evalState []`. Why not move that here?

File: src/Language/PureScript/Errors.hs [hunk 2]
```diff
@@ -250,12 +251,70 @@ errorMessage err = MultipleErrors [err]
...
+                                          then Just count
+                                          else findIndexInListHelp (count+1) f t
+replaceUnknowns' other = return other
+
+
+replaceUnknowns :: Type -> State [Unknown] Type
+replaceUnknowns = everywhereOnTypesTopDownM replaceUnknowns'
+
+
+simplifyUnknowns :: ErrorMessage -> State [Unknown] ErrorMessage
```
Comment (paf31) [inline src/Language/PureScript/Errors.hs]: This function looks useful, but can be far more generic. See if you can parameterize over `f`, then remove the type signature and see what `ghc` infers.

File: src/Language/PureScript/Errors.hs [hunk 3]
```diff
@@ -250,12 +251,70 @@ errorMessage err = MultipleErrors [err]
...
+      findIndexInList = findIndexInListHelp 0
+
+      findIndexInListHelp _ _ [] = Nothing
+      findIndexInListHelp count f (h:t) = if (f h)
+                                          then Just count
+                                          else findIndexInListHelp (count+1) f t
+replaceUnknowns' other = return other
+
+
+replaceUnknowns :: Type -> State [Unknown] Type
```
Comment (paf31) [inline src/Language/PureScript/Errors.hs]: Maybe stick `replaceUnknowns'` in a `where` clause?

File: src/Language/PureScript/Errors.hs [hunk 4]
```diff
@@ -250,12 +251,70 @@ errorMessage err = MultipleErrors [err]
 onErrorMessages :: (ErrorMessage -> ErrorMessage) -> MultipleErrors -> MultipleErrors
 onErrorMessages f = MultipleErrors . map f . runMultipleErrors
 
+
+replaceUnknowns' :: Type -> State [Unknown] Type
+replaceUnknowns' (TUnknown u) = do
+  table <- get
+  (newTable, newPos) <- return (lookupTable u table)
```
Comment (paf31) [inline src/Language/PureScript/Errors.hs]: maybe use `state` here, instead of `get` .. compute .. `put`

File: src/Language/PureScript/Errors.hs [hunk 5]
```diff
@@ -250,12 +251,70 @@ errorMessage err = MultipleErrors [err]
...
+  put newTable
+  return (TUnknown newPos)
+    where
+      lookupTable :: Unknown -> [Unknown] -> ([Unknown], Int)
+      lookupTable x table = case (findIndexInList (== x) table) of
+                                Nothing -> (table ++ [x], length table)
+                                Just i  -> (table, i)
+
+      findIndexInList :: (a -> Bool) -> [a] -> Maybe Int
```
Comment (paf31) [inline src/Language/PureScript/Errors.hs]: Try using `findIndex`:

https://hackage.haskell.org/package/base-4.8.0.0/docs/Data-List.html#v:findIndex

File: src/Language/PureScript/Errors.hs [hunk 6]
```diff
@@ -250,12 +252,55 @@ errorMessage err = MultipleErrors [err]
...
+    lookupTable x table = case (findIndex (== x) table) of
+                               Nothing -> (length table, table ++ [x])
+                               Just i  -> (i, table)
+
+    replaceUnknowns' :: Type -> State [Unknown] Type
+    replaceUnknowns' (TUnknown u) = state $ first TUnknown . lookupTable u
+    replaceUnknowns' other = return other
+
+
+onTypesInErrorMessageM :: (Applicative m) => (Type -> m Type) -> ErrorMessage -> m ErrorMessage
```
Comment (paf31) [inline src/Language/PureScript/Errors.hs]: Nice!

**Comment by paf31** [comment]
I think this looks great. Could you please just post an example so I can see what the errors look like now?

**Comment by balajirrao** [comment]
Ok, here's one.

``` haskell
module Unify where
       f x = x + 2    
       g x = x || true

       h = f >>> g
```
**New Error Message**

I just observed a couple of things - 'dict Prelude.Semigroupoid u698' is still shown the old way. I'll try to fix that.

Secondly, the unknown indices are odd numbers (u1, u3, u5..).  I can't figure out why.. It could happen if the simplification happens twice..

```
  Error applying function
      Prelude.(>>>)(<<dict Prelude.Semigroupoid u698>>)(Unify.f)
    of type
      u1 u3 u5 -> u1 u7 u5
    to argument
      Unify.g
    Error checking that value
      Unify.g
    has type
      Prim.Number -> u5
    Error checking that type
      Prim.Boolean -> Prim.Boolean
    subsumes type
      Prim.Number -> u5
    Error checking that type
      Prim.Number
    subsumes type
      Prim.Boolean
    Error unifying type
      Prim.Number
    with type
      Prim.Boolean
    Cannot unify type
      Prim.Number
    with type
      Prim.Boolean
```
**Original Error Message**
```
 Error applying function
      Prelude.(>>>)(<<dict Prelude.Semigroupoid u698>>)(Unify.f)
    of type
      u698 u696 u695 -> u698 u697 u695
    to argument
      Unify.g
    Error checking that value
      Unify.g
    has type
      Prim.Number -> u695
    Error checking that type
      Prim.Boolean -> Prim.Boolean
    subsumes type
      Prim.Number -> u695
    Error checking that type
      Prim.Number
    subsumes type
      Prim.Boolean
    Error unifying type
      Prim.Number
    with type
      Prim.Boolean
    Cannot unify type
      Prim.Number
    with type
      Prim.Boolean
```

**Comment by paf31** [comment]
Nice! Let's try to identify and fix that odd number issue though.

Let's worry about the dictionary thing a little later.

Another thing to think about is skolemized types. They are also numbered, so we should try to rename those too. You should be able to do it in the same pass - just look for the `Skolem` constructor.

File: src/Language/PureScript/Errors.hs [hunk 7]
```diff
@@ -250,12 +252,55 @@ errorMessage err = MultipleErrors [err]
 onErrorMessages :: (ErrorMessage -> ErrorMessage) -> MultipleErrors -> MultipleErrors
 onErrorMessages f = MultipleErrors . map f . runMultipleErrors
 
+replaceUnknowns :: Type -> State [Unknown] Type
+replaceUnknowns = everywhereOnTypesTopDownM replaceUnknowns'
+  where
+    lookupTable :: Unknown -> [Unknown] -> (Unknown, [Unknown])
+    lookupTable x table = case (findIndex (== x) table) of
```
Comment (paf31) [inline src/Language/PureScript/Errors.hs]: Actually `elemIndex` is the one you want, sorry.

Also, redundant parens here.

**Comment by balajirrao** [comment]
Ok, I'm on it.. My suspicion is right. The replacement is happening twice. Let me try to understand why..

**Comment by paf31** [comment]
Well replacement _should_ be idempotent. That would be one good way to test this.

**Comment by paf31** [comment]
Actually, don't worry about skolems for now. I'm going to handle them differently soon anyway.

File: src/Language/PureScript/Errors.hs [hunk 8]
```diff
@@ -514,32 +559,37 @@ prettyPrintSingleError full e = prettyPrintErrorMessage (if full then e else sim
...
     wrap (Just pos) = PositionedError pos
 
+
 -- |
 -- Pretty print multiple errors
 --
 prettyPrintMultipleErrors :: Bool -> MultipleErrors -> String
-prettyPrintMultipleErrors = prettyPrintMultipleErrorsWith "Error:" "Multiple errors:"
-    
+prettyPrintMultipleErrors = curry $ flip evalState [] . (uncurry $ prettyPrintMultipleErrorsWith "Error:" "Multiple errors:")
```
Comment (paf31) [inline src/Language/PureScript/Errors.hs]: I find this pointfree-style code a little hard to read. Can we just keep the `full` argument as a named argument?

**Comment by balajirrao** [comment]
I now get the following error (on a slightly different program).

```
Error applying function
      Prelude.(>>>)(<<dict Prelude.Semigroupoid u701>>)(Unify.g)
    of type
      u0 u1 u2 -> u0 u3 u2
    to argument
      true
    Error checking that value
      true
    has type
      Prim.Boolean -> u2
    Expression
      true
    does not have type
      Prim.Boolean -> u2
```

**Comment by paf31** [comment]
Looks great! Would you say this is ready to merge now? If so, could you please squash your commits and I'll merge this in:

```
git rebase -i purescript/master
...
git push origin/branch --force
```

**Comment by balajirrao** [comment]
Thanks! I think it's ready to be merged. I'll squash the commits and push.

**Comment by paf31** [comment]
:+1: Thanks!

**Pull request merged**